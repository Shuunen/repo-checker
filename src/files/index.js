const CheckConfigs = require('./configs')
const CheckReadme = require('./readme')
const CheckPackage = require('./package')
const CheckLicense = require('./license')
const CheckTravis = require('./travis')
const CheckRenovate = require('./renovate')
const CheckEditorConfig = require('./editor-config')
const CheckNvmRc = require('./nvmrc')

module.exports = [CheckConfigs, CheckNvmRc, CheckReadme, CheckPackage, CheckLicense, CheckTravis, CheckRenovate, CheckEditorConfig]
