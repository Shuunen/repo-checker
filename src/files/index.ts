export * from './configs'
export * from './editor-config'
export * from './eslint'
export * from './gh-workflow'
export * from './license'
export * from './nvmrc'
export * from './package'
export * from './readme'
export * from './renovate'
export * from './repo-checker'
export * from './travis'
// dam lint
