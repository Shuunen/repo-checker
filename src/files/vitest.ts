import { FileBase } from '../file.ts'

// eslint-disable-next-line no-export
export class VitestFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  // eslint-disable-next-line max-lines-per-function
  public async start() {
    if (!this.data.isUsingVitest) return
    const hasMtsFile = await this.fileExists('vitest.config.mts')
    const hasJsFile = await this.fileExists('vitest.config.js')
    const hasTsFile = await this.fileExists('vitest.config.ts')
    const hasMjsFile = await this.fileExists('vitest.config.mjs')
    if (hasMtsFile) await this.inspectFile('vitest.config.mts')
    /* c8 ignore next 3 */
    if (hasJsFile) await this.inspectFile('vitest.config.js')
    if (hasTsFile) await this.inspectFile('vitest.config.ts')
    if (hasMjsFile) await this.inspectFile('vitest.config.mjs')
    this.couldContains('a pool thread setting', /pool: 'threads'/u)
  }
}
