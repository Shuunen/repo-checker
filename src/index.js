#!/usr/bin/env node

const arg = require('arg')
const path = require('path')

const check = require('./check')
const log = require('./logger')

function init () {
  const args = arg({ '--target': String, '--data': String, '--fix': Boolean }, { argv: process.argv.slice(2) })
  let target = args['--target']
  const doFix = args['--fix']
  if (!target) {
    log.info('\n', `no target specified via : --target=path/to/directory`)
    log.info(`targeting current directory...`, '\n')
    target = '.'
  }
  let data = {}
  if (args['--data']) {
    const p = path.join(__dirname, '..', args['--data'])
    log.info('loading data from', p)
    data = require(p)
  } else if (doFix) {
    log.warn('you should provide data to enhance fix')
  }
  log.start()
    .then(() => check(target, data, doFix))
    .catch(err => log.error(err))
    .then(() => log.end())
}

init()
