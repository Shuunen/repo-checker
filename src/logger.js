/* eslint-disable no-console */
const fs = require('fs')
const chalk = require('chalk')

const pkg = require('../package.json')

class Logger {
  get indent () {
    return '    '.repeat(this.indentLevel)
  }

  get date () {
    return new Date().toISOString().replace('T', ' ').split('.')[0]
  }

  constructor (filePath) {
    this.file = fs.createWriteStream(filePath, { flags: 'a' })
    this.indentLevel = 0
  }

  async _write (...args) {
    this.file.write(args.join(' ') + '\n')
  }

  async setIndentLevel (level) {
    this.indentLevel = level
  }

  debug (...args) {
    return this._write(this.indent, 'üîπ', ...args)
  }

  info (...args) {
    console.log(this.indent, '‚¨ú ', ...args)
    return this._write(this.indent, '‚¨ú', ...args)
  }

  error (...args) {
    console.error(chalk.redBright(this.indent, '‚ùå ', ...args))
    return this._write(this.indent, '‚ùå', ...args)
  }

  warn (...args) {
    console.log(chalk.yellowBright(this.indent, '‚ö†Ô∏è ', ...args))
    return this._write(this.indent, '‚ö†Ô∏è', ...args)
  }

  success (...args) {
    console.log(chalk.green(this.indent, '‚úîÔ∏è ', ...args))
    return this._write(this.indent, '‚úîÔ∏è', ...args)
  }

  test (ok, msg, justWarn) {
    if (ok) {
      this.success(msg)
    } else if (justWarn) {
      this.warn(msg)
    } else {
      this.error(msg)
    }
  }

  fix (...args) {
    console.log(chalk.blueBright(this.indent, '‚¨ú ', ...args))
    return this._write(this.indent, '‚¨ú', ...args)
  }

  line () {
    console.log('')
    return this._write('')
  }

  start () {
    this.line()
    this._write(`‚¨áÔ∏è--- Entry from ${this.date} ---‚¨áÔ∏è`)
    this.info(`${pkg.name} v${pkg.version} is starting`)
    return this.line()
  }

  end () {
    this.line()
    this.info(pkg.name, 'has finished')
    return this.line()
  }
}

const log = new Logger(pkg.config.logFile)

module.exports = log
