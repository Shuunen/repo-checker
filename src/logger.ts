import { bgBlue, black, Logger } from 'shuutils'

class ExtendedLogger extends Logger {
  public fix (...stuff: readonly string[]): void {
    const items = Array.from(stuff)
    items.push(bgBlue(black('[ fixed ]')))
    this.info(...items)
  }

  public unknownError (error: unknown): void {
    if (error instanceof Error) { this.error(error.message); return }
    if (typeof error === 'string') { this.error(error); return }
    this.error('Unknown error format, could not log it')
  }
}

export const log = new ExtendedLogger()
