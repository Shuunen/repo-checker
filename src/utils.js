const fs = require('fs')
const path = require('path')
const log = require('./logger')
const Git = require('simple-git/promise')

async function isGitFolder (folderPath) {
  if (!fs.statSync(folderPath).isDirectory()) {
    return false
  }
  return Git(folderPath).checkIsRepo().then(isRepo => {
    log.debug(`${isRepo ? 'is' : 'NOT'} a git repo : ${folderPath}`)
    return isRepo
  }).catch(err => {
    log.error(`failed at testing git repo : ${folderPath}`, err)
    throw err
  })
}

async function getGitFolders (folderPath) {
  if (await isGitFolder(folderPath)) {
    return [folderPath]
  }
  return new Promise((resolve, reject) => {
    fs.readdir(folderPath, async (err, filePaths) => {
      if (err) {
        log.error(`failed at reading dir : ${folderPath}`, err)
        reject(err)
      }
      let gitDirectories = []
      for (const filePath of filePaths) {
        const p = path.join(folderPath, filePath)
        if (await isGitFolder(p)) {
          gitDirectories.push(p)
        }
      }
      resolve(gitDirectories)
    })
  })
}

function augmentData (folderPath, dataSource) {
  const data = Object.assign({}, dataSource)
  return Git(folderPath).getRemotes(true).then(remotes => {
    const matches = JSON.stringify(remotes).match(/([\w-]+)\/([\w-]+)\.git/)
    if (matches.length === 3) {
      data.user_id = matches[1]
      data.user_id_lowercase = data.user_id.toLowerCase()
      data.repo_id = matches[2]
    }
    return data
  })
}

function folderContainsFile (folderPath, fileName) {
  if (!folderPath || !fileName) {
    return log.error('folderContainsFile miss arguments')
  }
  return new Promise((resolve, reject) => {
    fs.access(path.join(folderPath, fileName), fs.F_OK, (err) => {
      if (err) {
        resolve(false)
      }
      resolve(true)
    })
  })
}

function createFile (folderPath, fileName, fileContent) {
  return new Promise((resolve, reject) => {
    fs.writeFile(path.join(folderPath, fileName), fileContent, 'utf8', (err) => {
      if (err) {
        log.error(err)
        resolve(false)
      }
      resolve(true)
    })
  })
}

function readFile (folderPath, fileName, returnEmptyIfNotExists) {
  return new Promise((resolve, reject) => {
    fs.readFile(path.join(folderPath, fileName), 'utf8', (err, content) => {
      if (err) {
        if (returnEmptyIfNotExists) {
          resolve('')
        }
        reject(err)
      }
      resolve(content)
    })
  })
}

function fillTemplate (template, data) {
  let str = (typeof template === 'object' ? JSON.stringify(template, null, 2) : template) || ''
  if (!str) {
    return str
  }
  const tokens = str.match(/\{{\s?([^}\s]+)\s?}\}/g)
  if (!tokens) {
    return str
  }
  return tokens.reduce(
    (acc, token) => {
      const key = token.replace(/[{\s}]/g, '')
      let value = data && data[key]
      if (!value) {
        log.error(`please provide variable "${key}"`)
      }
      acc = acc.replace(token, value || '')
      return acc
    },
    str
  )
}

module.exports = { log, getGitFolders, fillTemplate, folderContainsFile, createFile, readFile, augmentData }
