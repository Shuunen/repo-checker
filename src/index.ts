#!/usr/bin/env node

import { Nb } from 'shuutils'
import { log } from './logger'
import { start } from './main'

// eslint-disable-next-line unicorn/prefer-top-level-await
start().catch((error: unknown) => {
  log.unknownError(error)
  process.exit(Nb.One)
})
