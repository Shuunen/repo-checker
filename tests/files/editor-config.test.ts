import { expect, it } from 'vitest'
import { EditorConfigFile } from '../../src/files'
import { log } from '../../src/logger'
import { cleanInstanceForSnap, promiseFalse, promiseTrue, promiseVoid } from '../utils'

it('editor config A missing file', async () => {
  log.disable()
  const instance = new EditorConfigFile()
  instance.checkFileExists = promiseFalse
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})

it('editor config B file', async () => {
  log.disable()
  const instance = new EditorConfigFile()
  instance.checkFileExists = promiseTrue
  instance.inspectFile = promiseVoid
  instance.fileContent = ''
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})
