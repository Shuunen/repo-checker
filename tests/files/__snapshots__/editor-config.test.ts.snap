// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`editor config A missing file 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": false,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [],
}
`;

exports[`editor config B file 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": false,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [
    "does-not-have-space-indent-indent-style-space",
    "does-not-have-indent-size-of-2-indent-size-2",
    "does-not-have-unix-style-line-endings-end-of-line-lf",
    "does-not-have-utf-8-encoding-charset-utf-8",
  ],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [
    "has-no-specific-html-indent-rule",
  ],
  "warnings": [
    "could-have-whitespace-trailing-trim-trailing-whitespace-true",
    "could-have-final-new-line-rule-insert-final-newline-true",
  ],
}
`;
